File: src/app/repository/ArticleRepository.java
package app.repository;

import app.model.Article;

import java.sql.*;
import java.util.ArrayList;
import java.util.List;

public final class ArticleRepository extends _BaseRepository {
  private ArticleRepository() {
  }

  public static Article getArticle(int articleId) {
    final String SQL = "SELECT * from `ArticleCounted` where `articleId`=?";
    return connect(connection -> {
      final PreparedStatement statement = connection.prepareStatement(SQL);
      statement.setInt(1, articleId);
      final ResultSet res = statement.executeQuery();
      return res.next() ? Article.from(res) : null;
    });
  }

  public static List<Article> getUserArticles(int userId) {
    final String SQL = "SELECT * from `ArticleCounted` where `writer_userId`=?";
    return connect(connection -> {
      final PreparedStatement statement = connection.prepareStatement(SQL);
      statement.setInt(1, userId);
      final ResultSet res = statement.executeQuery();
      final List<Article> result = new ArrayList<>();
      while (res.next())
        result.add(Article.from(res));
      return result;
    });
  }

  public static Article saveArticle(Article article) {
    final String SQL = "INSERT OR REPLACE INTO `Article` (`articleId`, `writer_userId`, `title`, `content`, `time`, `featured`) VALUES (?,?,?,?,?,?) RETURNING *;";
    return connect(connection -> {
      final PreparedStatement statement = connection.prepareStatement(SQL);

      if (article.articleId > 0)
        statement.setInt(1, article.articleId);
      else
        statement.setNull(1, Types.INTEGER);
      statement.setInt(2, article.writerUserId);
      statement.setString(3, article.title);
      statement.setString(4, article.content);
      if (article.time != null)
        statement.setLong(5, article.time.getTime());
      else
        statement.setLong(5, System.currentTimeMillis());
      statement.setInt(6, article.featured ? 1 : 0);

      final ResultSet res = statement.executeQuery();
      return res.next() ? Article.from(res) : null;
    });
  }

  public static Article deleteArticle(int articleId) {
    final String SQL = "DELETE FROM `Article` WHERE `articleId`=? RETURNING *;";
    return connect(connection -> {
      final PreparedStatement statement = connection.prepareStatement(SQL);
      statement.setInt(1, articleId);
      final ResultSet res = statement.executeQuery();
      return res.next() ? Article.from(res) : null;
    });
  }

  public static List<Article> getUserHome(int userId) {
    final String SQL = "SELECT * from `Article` where `articleId` in (select `articleId` from Home where `userId`=? order by `time` desc)";
    return connect(connection -> {
      final PreparedStatement statement = connection.prepareStatement(SQL);
      statement.setInt(1, userId);
      final ResultSet res = statement.executeQuery();
      final List<Article> result = new ArrayList<>();
      while (res.next())
        result.add(Article.from(res));
      return result;
    });
  }
}

File: src/app/repository/ArticleRepository.kt
@file:JvmName("Repository")
@file:JvmMultifileClass

package app.repository

import app.model.Article

fun getUserArticle(homeUserId: Int, articleId: Int): Article? =
	connect {
		val SQL = """
			SELECT AST.*, HA.home_userId , HA.home_count, HA.home_time, HA.home_isLiked
			from ArticleStat AST
			LEFT JOIN HomeArticle HA ON AST.articleId= HA.articleId AND HA.home_userId=?
			where AST.articleId=?;
		""".trimIndent()
		val statement = it.prepareStatement(SQL)
		statement.setInt(1, homeUserId)
		statement.setInt(2, articleId)
		statement.executeQuery()
			.singleOf<Article>()
	}

fun listUserArticles(homeUserId: Int): List<Article> =
	connect {
		val SQL = """
			SELECT * from HomeArticle where home_userId=? AND writer_userId=?;
		""".trimIndent()
		val statement = it.prepareStatement(SQL)
		statement.setInt(1, homeUserId)
		statement.setInt(2, homeUserId)
		statement.executeQuery()
			.listOf<Article>()
	} ?: emptyList()

fun listHomeUserArticles(homeUserId: Int): List<Article> =
	connect {
		val SQL = """
			SELECT * from HomeArticle where home_userId=?;
		""".trimIndent()
		val statement = it.prepareStatement(SQL)
		statement.setInt(1, homeUserId)
		statement.executeQuery()
			.listOf<Article>()
	} ?: emptyList()

fun saveArticle(article: Article): Article? {
	val articleId = connect {
		val SQL_INSERT = """
			INSERT INTO Article (articleId, title, content, time, featured, writer_userId)
			VALUES (NULL,?,?,?,?,?) RETURNING articleId;
	 """.trimIndent()
		val SQL_UPDATE = """
			UPDATE Article SET title=?, content=?, time=?, featured=?
			WHERE writer_userId=? AND articleId=? RETURNING articleId;
	 """.trimIndent()

		val stmt = it.prepareStatement(if (article.articleId > 0) SQL_UPDATE else SQL_INSERT)
		stmt.setString(1, article.title)
		stmt.setString(2, article.content)
		stmt.setLong(3, System.currentTimeMillis())
		stmt.setInt(4, if (article.featured) 1 else 0)
		stmt.setInt(5, article.writerUserId)
		if (article.articleId > 0)
			stmt.setInt(6, article.articleId)

		stmt.executeQuery()
			.singleOf<Int>()
	} ?: return null

	return getUserArticle(article.writerUserId, articleId)
}

fun deleteArticle(homeUserId: Int, articleId: Int): Boolean =
	connect {
		val SQL = """
			DELETE FROM Article WHERE articleId=? AND writer_userId=?;
		""".trimIndent()
		val statement = it.prepareStatement(SQL)
		statement.setInt(1, articleId)
		statement.setInt(2, homeUserId)
		statement.executeUpdate() > 0
	} == true
